File: src/erouter/core/seed.py
"""Seed subgraph for column generation (spec §5.3).

The union of:

* the `k` shortest `src -> dst` paths under edge length `eps_p`,
* the top-conductance arcs incident to any node those paths touch (breadth,
  not depth -- the solver finds depth by itself),
* any negative-`eps` cycle reachable from the path set, which is free
  arbitrage the router should absorb.

`eps_p` can be negative -- a favourably dislocated pool is an EMF -- so this
uses SPFA (queue-based Bellman-Ford) rather than Dijkstra, and detects negative
cycles instead of looping on them.

**Seed quality only affects how many column-generation rounds run, never the
answer** (§5.5's certificate is what guarantees correctness), so this file is
allowed to be approximate and is not allowed to be slow.
"""

from __future__ import annotations

import heapq
import os
from collections import deque
from dataclasses import dataclass, field

import numpy as np

from . import accel as _accel
from .graph import ArcArrays

INF = float("inf")

# Relaxation depth for the seed searches.  Real Curve routes are 1-3 hops
# (mainnet universe: mean 1.97, max 3), so this is generous -- but it turns each
# search from "iterate to a fixed point over 301 nodes" into a bounded sweep.
# Safe in a way bounding the solve would not be: §5.5's certificate prices out
# *all* m arcs, so a path the seed misses is still found by column generation.
MAX_HOPS = 8
#: Opt-in on the same switch as the rest of the port.
_ACCEL_ON = os.environ.get("EROUTER_ACCEL", "") == "1"


@dataclass(slots=True)
class Adjacency:
    """Outgoing arc indices per node, in CSR form.

    Carries plain-`list` mirrors beside the numpy arrays: the relaxation loop
    below reads single elements ~300k times per route, and a boxed numpy scalar
    costs 72.8 ns against 21.2 ns for a list element.  Building the mirrors is
    one vectorised pass; the loop that consumes them is the hottest in the
    router.
    """

    starts: np.ndarray
    arcs: np.ndarray
    starts_list: list[int] = field(default_factory=list)
    arcs_list: list[int] = field(default_factory=list)

    def out(self, node: int) -> np.ndarray:
        return self.arcs[self.starts[node] : self.starts[node + 1]]


def build_adjacency(tau: np.ndarray, sig: np.ndarray, n: int) -> Adjacency:
    del sig
    order = np.argsort(tau, kind="stable")
    counts = np.bincount(tau, minlength=n)
    starts = np.zeros(n + 1, dtype=np.int64)
    np.cumsum(counts, out=starts[1:])
    arcs = order.astype(np.int64)
    return Adjacency(starts, arcs, starts.tolist(), arcs.tolist())


@dataclass(slots=True)
class ShortestPath:
    arcs: list[int] = field(default_factory=list)
    length: float = INF
    found: bool = False
    negative_cycle: list[int] = field(default_factory=list)


def spfa(
    g: ArcArrays,
    src: int,
    dst: int,
    adj: Adjacency,
    *,
    banned_arcs: set[int] | None = None,
    banned_nodes: set[int] | None = None,
    weights: np.ndarray | None = None,
    max_hops: int = MAX_HOPS,
) -> ShortestPath:
    """Shortest `src -> dst` path by arc length `eps`, tolerating negative arcs.

    Bellman-Ford with a FIFO queue (SPFA), not Dijkstra: `eps_p` can be
    **negative** -- a favourably dislocated pool is an EMF -- and Dijkstra needs
    non-negative weights.  Returns the arc indices of the path; a reachable
    negative cycle comes back separately rather than the search diverging.

    `max_hops` caps the path depth, which makes the search approximate (a node
    reached cheaply but deep may keep a shallower, dearer label).  §5.3 permits
    that: the seed decides how many CG rounds run, not what the answer is.

    Everything in the loop is a plain Python `int`/`float` read out of a list;
    the numpy arrays cost 3.4x as much per access.
    """
    # The compiled search when installed: same algorithm, same tie-breaks, and
    # `tests/test_seed_differential.py` differs the two.
    if _ACCEL_ON and _accel.available():
        got = _accel.shortest_path(
            g, src, dst, banned_arcs=banned_arcs, banned_nodes=banned_nodes,
            weights=weights, max_hops=max_hops)
        if got is not None:
            if got["negative_cycle"]:
                return ShortestPath(negative_cycle=list(got["negative_cycle"]))
            if not got["found"]:
                return ShortestPath()
            return ShortestPath(list(got["arcs"]), float(got["length"]), True)

    n = g.n_nodes
    banned_arcs = banned_arcs or set()
    banned_nodes = banned_nodes or set()
    cost = (g.eps if weights is None else weights).tolist()
    head_of = g.sig.tolist()
    starts, arc_order = adj.starts_list, adj.arcs_list

    dist = [INF] * n
    parent = [-1] * n  # arc index used to reach the node
    hops = [0] * n
    in_queue = [False] * n
    # Depth-bounding already forces termination -- every relaxation strictly
    # lowers a `dist` and only walks of at most `max_hops` arcs exist -- so this
    # is a backstop against a pathology in the bound, not the mechanism.
    budget = 8 * max_hops * n

    dist[src] = 0.0
    # FIFO through a deque: `list.pop(0)` memmoves the queue, ~127k times a route.
    queue: deque[int] = deque((src,))
    in_queue[src] = True

    while queue and budget > 0:
        node = queue.popleft()
        in_queue[node] = False
        depth = hops[node] + 1
        if depth > max_hops:
            continue
        base = dist[node]
        for k in range(starts[node], starts[node + 1]):
            arc = arc_order[k]
            if arc in banned_arcs:
                continue
            head = head_of[arc]
            if head in banned_nodes:
                continue
            candidate = base + cost[arc]
            if candidate < dist[head] - 1e-15:
                dist[head] = candidate
                parent[head] = arc
                hops[head] = depth
                budget -= 1
                if not in_queue[head]:
                    queue.append(head)
                    in_queue[head] = True

    if dist[dst] == INF:
        return ShortestPath()

    # Walk the parent pointers back.  A negative cycle shows up here and only
    # here: `dist` keeps falling around the loop, so the chain from `dst`
    # re-enters a node it already passed instead of reaching `src`.  Detecting
    # it on the walk needs no relaxation counters in the inner loop.
    tail_of = g.tau.tolist()
    arcs: list[int] = []
    seen_at: dict[int, int] = {}
    node = dst
    while node != src:
        if node in seen_at:
            cycle = arcs[seen_at[node] :]
            cycle.reverse()
            return ShortestPath(negative_cycle=cycle)
        seen_at[node] = len(arcs)
        arc = parent[node]
        if arc < 0:
            return ShortestPath()
        arcs.append(arc)
        node = tail_of[arc]
    arcs.reverse()
    return ShortestPath(arcs, float(dist[dst]), True)


def k_shortest_paths(
    g: ArcArrays, src: int, dst: int, k: int = 10, adj: Adjacency | None = None
) -> list[list[int]]:
    """Yen's algorithm over `eps`, returning arc-index paths.

    Only ever returns genuine `src -> dst` paths.  A negative cycle is not one:
    conflating the two lets a caller warm-start the solver from a loop that
    does not contain `src`, which then reports the graph as disconnected.
    """
    adj = adj or build_adjacency(g.tau, g.sig, g.n_nodes)

    weights = None
    first = spfa(g, src, dst, adj)
    if first.negative_cycle:
        # §5.3: "shift by min(0, min_p eps_p) and correct".  Shifting changes a
        # path's cost per hop rather than preserving the order, so the result is
        # only approximately cheapest -- fine, since seed quality affects rounds.
        weights = g.eps - min(0.0, float(g.eps.min()))
        first = spfa(g, src, dst, adj, weights=weights)
    if not first.found:
        return []

    accepted: list[list[int]] = [first.arcs]
    candidates: list[tuple[float, list[int]]] = []
    seen = {tuple(first.arcs)}

    while len(accepted) < k:
        previous = accepted[-1]
        for i in range(len(previous)):
            root = previous[:i]
            spur_node = int(g.tau[previous[i]])

            banned_arcs = {
                path[i]
                for path in accepted
                if len(path) > i and path[:i] == root
            }
            banned_nodes = {int(g.tau[arc]) for arc in root}

            spur = spfa(
                g, spur_node, dst, adj, weights=weights,
                banned_arcs=banned_arcs, banned_nodes=banned_nodes,
            )
            if not spur.found:
                continue
            whole = root + spur.arcs
            key = tuple(whole)
            if key in seen:
                continue
            seen.add(key)
            heapq.heappush(candidates, (float(np.sum(g.eps[whole])), whole))

        if not candidates:
            break
        accepted.append(heapq.heappop(candidates)[1])

    return accepted


def seed_subgraph(
    g: ArcArrays,
    src: int,
    dst: int,
    *,
    k: int = 10,
    breadth: int = 8,
    include_negative: bool = True,
) -> np.ndarray:
    """Boolean mask of arcs to start column generation from.

    `breadth` is how many top-conductance arcs to add per touched node.  Breadth
    rather than depth: the solver discovers depth through pricing-out, but it can
    only price in an arc that some node makes reachable.
    """
    mask = np.zeros(g.m, bool)
    if g.m == 0:
        return mask

    adj = build_adjacency(g.tau, g.sig, g.n_nodes)
    paths = k_shortest_paths(g, src, dst, k, adj)
    for path in paths:
        mask[path] = True

    touched = {src, dst}
    for path in paths:
        for arc in path:
            touched.add(int(g.tau[arc]))
            touched.add(int(g.sig[arc]))

    # Top-G arcs incident to any touched node, in either direction.
    incident_to = np.isin(g.tau, list(touched)) | np.isin(g.sig, list(touched))
    for node in touched:
        local = np.flatnonzero(incident_to & ((g.tau == node) | (g.sig == node)))
        if local.size > breadth:
            local = local[np.argsort(-g.G[local])[:breadth]]
        mask[local] = True

    if include_negative:
        # Free arbitrage the router should absorb; also cheap insurance against
        # a seed that missed a battery path entirely.
        mask[g.eps < 0] = True

    return mask

File: rust/src/seed.rs
//! Shortest `src -> dst` path by arc length, tolerating negative arcs (§5.3).
//!
//! A port of `core/seed.py`'s `spfa`.  Bellman-Ford with a FIFO queue, not
//! Dijkstra: `eps_p` can be negative -- a favourably dislocated pool is an EMF
//! -- and Dijkstra needs non-negative weights.
//!
//! The Python is already careful (plain lists, a `deque`, no numpy in the
//! loop) and still costs 0.3 ms a call because the loop runs ~300k times a
//! route.  That is the shape a compiled port helps most: no clever
//! arithmetic, just an enormous number of very cheap steps.

/// Relaxation depth for the seed searches. Real Curve routes are 1-3 hops
/// (mainnet universe: mean 1.97, max 3), so this is generous -- but it turns
/// each search from "iterate to a fixed point over 301 nodes" into a bounded
/// sweep. Safe in a way bounding the solve would not be: §5.5's certificate
/// prices out *all* m arcs, so a path the seed misses is still found by column
/// generation.
pub const MAX_HOPS: usize = 8;

/// Where the walk back from `dst` ended up.
pub struct Path {
    pub arcs: Vec<usize>,
    pub length: f64,
    pub found: bool,
    /// Set instead of `arcs` when the parent chain re-enters a node: `dist`
    /// keeps falling around a loop rather than reaching `src`.
    pub negative_cycle: Vec<usize>,
}

impl Path {
    fn empty() -> Self {
        Path { arcs: Vec::new(), length: f64::INFINITY, found: false,
               negative_cycle: Vec::new() }
    }
}

/// CSR adjacency by tail node: `arcs[starts[v]..starts[v + 1]]` leave `v`.
pub struct Adjacency {
    pub starts: Vec<usize>,
    pub arcs: Vec<usize>,
}

pub fn build_adjacency(tau: &[i64], n_nodes: usize) -> Adjacency {
    let m = tau.len();
    let mut counts = vec![0usize; n_nodes + 1];
    for &t in tau {
        counts[t as usize + 1] += 1;
    }
    for v in 0..n_nodes {
        counts[v + 1] += counts[v];
    }
    let starts = counts.clone();
    let mut cursor = counts;
    let mut arcs = vec![0usize; m];
    for (p, &t) in tau.iter().enumerate() {
        let slot = cursor[t as usize];
        arcs[slot] = p;
        cursor[t as usize] += 1;
    }
    Adjacency { starts, arcs }
}

#[allow(clippy::too_many_arguments)]
pub fn spfa(
    tau: &[i64],
    sig: &[i64],
    cost: &[f64],
    n_nodes: usize,
    adj: &Adjacency,
    src: usize,
    dst: usize,
    banned_arcs: &[bool],
    banned_nodes: &[bool],
    max_hops: usize,
) -> Path {
    let n = n_nodes;
    let mut dist = vec![f64::INFINITY; n];
    let mut parent = vec![-1i64; n];
    let mut hops = vec![0usize; n];
    let mut in_queue = vec![false; n];

    // Depth-bounding already forces termination; this is a backstop against a
    // pathology in the bound itself, not the mechanism.
    let mut budget = (8 * max_hops * n) as i64;

    dist[src] = 0.0;
    let mut queue: std::collections::VecDeque<usize> = std::collections::VecDeque::new();
    queue.push_back(src);
    in_queue[src] = true;

    while let Some(node) = queue.pop_front() {
        if budget <= 0 {
            break;
        }
        in_queue[node] = false;
        let depth = hops[node] + 1;
        if depth > max_hops {
            continue;
        }
        let base = dist[node];
        for k in adj.starts[node]..adj.starts[node + 1] {
            let arc = adj.arcs[k];
            if banned_arcs[arc] {
                continue;
            }
            let head = sig[arc] as usize;
            if banned_nodes[head] {
                continue;
            }
            let candidate = base + cost[arc];
            if candidate < dist[head] - 1e-15 {
                dist[head] = candidate;
                parent[head] = arc as i64;
                hops[head] = depth;
                budget -= 1;
                if !in_queue[head] {
                    queue.push_back(head);
                    in_queue[head] = true;
                }
            }
        }
    }

    if !dist[dst].is_finite() {
        return Path::empty();
    }

    // Walk the parent pointers back.  A negative cycle shows up here and only
    // here, and detecting it on the walk costs nothing.
    let mut arcs: Vec<usize> = Vec::new();
    let mut seen_at = vec![usize::MAX; n];
    let mut node = dst;
    while node != src {
        if seen_at[node] != usize::MAX {
            let mut cycle: Vec<usize> = arcs[seen_at[node]..].to_vec();
            cycle.reverse();
            return Path { arcs: Vec::new(), length: f64::INFINITY, found: false,
                          negative_cycle: cycle };
        }
        seen_at[node] = arcs.len();
        let arc = parent[node];
        if arc < 0 {
            return Path::empty();
        }
        arcs.push(arc as usize);
        node = tau[arc as usize] as usize;
    }
    arcs.reverse();
    let length = dist[dst];
    Path { arcs, length, found: true, negative_cycle: Vec::new() }
}

/// Yen's algorithm over `eps`, returning arc-index paths.
///
/// Only ever returns genuine `src -> dst` paths. A negative cycle is not one:
/// conflating the two lets a caller warm-start the solver from a loop that
/// does not contain `src`, which then reports the graph as disconnected.
pub fn k_shortest_paths(
    tau: &[i64], sig: &[i64], eps: &[f64], n_nodes: usize,
    src: usize, dst: usize, k: usize, adj: &Adjacency,
) -> Vec<Vec<usize>> {
    let m = tau.len();
    let none = vec![false; m];
    let no_nodes = vec![false; n_nodes];
    let run = |cost: &[f64], from: usize, banned_arcs: &[bool], banned_nodes: &[bool]| {
        spfa(tau, sig, cost, n_nodes, adj, from, dst, banned_arcs, banned_nodes, MAX_HOPS)
    };

    let mut first = run(eps, src, &none, &no_nodes);
    // §5.3: "shift by min(0, min_p eps_p) and correct". Shifting changes a
    // path's cost per hop rather than preserving the order, so the result is
    // only approximately cheapest -- fine, since seed quality affects rounds.
    let shifted: Option<Vec<f64>> = if first.negative_cycle.is_empty() {
        None
    } else {
        let floor = eps.iter().copied().fold(0.0f64, f64::min);
        let weights: Vec<f64> = eps.iter().map(|v| v - floor).collect();
        first = run(&weights, src, &none, &no_nodes);
        Some(weights)
    };
    let cost: &[f64] = shifted.as_deref().unwrap_or(eps);
    if !first.found {
        return Vec::new();
    }

    let mut accepted: Vec<Vec<usize>> = vec![first.arcs.clone()];
    // Yen's B-list. A `BinaryHeap` would need the reverse of Python's ordering
    // and its tie-break is not specified; this is a handful of entries and the
    // pop is the same `(cost, path)` minimum `heapq` takes -- lexicographic on
    // the path where the costs tie, which decides which of two equal-length
    // routes reaches the ballot.
    let mut candidates: Vec<(f64, Vec<usize>)> = Vec::new();
    let mut seen: Vec<Vec<usize>> = vec![first.arcs];

    while accepted.len() < k {
        let previous = accepted[accepted.len() - 1].clone();
        for i in 0..previous.len() {
            let root = &previous[..i];
            let spur_node = tau[previous[i]] as usize;

            let mut banned_arcs = vec![false; m];
            for path in &accepted {
                if path.len() > i && &path[..i] == root {
                    banned_arcs[path[i]] = true;
                }
            }
            let mut banned_nodes = vec![false; n_nodes];
            for &arc in root {
                banned_nodes[tau[arc] as usize] = true;
            }

            let spur = run(cost, spur_node, &banned_arcs, &banned_nodes);
            if !spur.found {
                continue;
            }
            let mut whole = root.to_vec();
            whole.extend(spur.arcs);
            if seen.contains(&whole) {
                continue;
            }
            seen.push(whole.clone());
            // Scored on `eps`, not on the shifted weights: the shift is only
            // there to make the search terminate.
            let length: f64 = whole.iter().map(|&a| eps[a]).sum();
            candidates.push((length, whole));
        }

        if candidates.is_empty() {
            break;
        }
        let mut best = 0usize;
        for at in 1..candidates.len() {
            if less(&candidates[at], &candidates[best]) {
                best = at;
            }
        }
        accepted.push(candidates.remove(best).1);
    }

    accepted
}

/// `(cost, path) < (cost, path)` the way a tuple compares in Python: the
/// number first, then the list, element by element and then by length.
fn less(a: &(f64, Vec<usize>), b: &(f64, Vec<usize>)) -> bool {
    if a.0 != b.0 {
        return a.0 < b.0;
    }
    a.1 < b.1
}

#[cfg(test)]
mod tests {
    use super::*;

    fn run(tau: &[i64], sig: &[i64], cost: &[f64], n: usize, src: usize, dst: usize) -> Path {
        let adj = build_adjacency(tau, n);
        spfa(tau, sig, cost, n, &adj, src, dst,
             &vec![false; tau.len()], &vec![false; n], 8)
    }

    #[test]
    fn it_takes_the_cheaper_of_two_lanes() {
        let got = run(&[0, 0], &[1, 1], &[2.0, 1.0], 2, 0, 1);
        assert!(got.found);
        assert_eq!(got.arcs, vec![1]);
        assert!((got.length - 1.0).abs() < 1e-15);
    }

    #[test]
    fn it_chains_hops() {
        let got = run(&[0, 1], &[1, 2], &[1.0, 1.0], 3, 0, 2);
        assert_eq!(got.arcs, vec![0, 1]);
    }

    #[test]
    fn a_negative_arc_is_taken_rather_than_diverging() {
        let got = run(&[0, 0, 1], &[2, 1, 2], &[0.0, -1.0, -1.0], 3, 0, 2);
        assert!(got.found, "dijkstra would have stopped at the zero-cost arc");
        assert_eq!(got.arcs, vec![1, 2]);
    }

    #[test]
    fn an_unreachable_target_is_not_found() {
        let got = run(&[0], &[1], &[1.0], 3, 0, 2);
        assert!(!got.found && got.negative_cycle.is_empty());
    }
}
